### ISSUE TITLE
High PostgreSQL CPU Usage with long running Query

### ISSUE THREAD

**Issue by ravermeister** [owncloud/music#997]
Hi, I noticed that the Query that is used to Cleanup the Artists here:
https://github.com/owncloud/music/blob/8714acbbc3a2ecd0b6981374d2751b4d7876c8b6/lib/Db/Maintenance.php#L136
results in 100% CPU Usage from PostgreSQL and the Query takes a very long time (longer than the cleanup schedule time).

After some searching in the Web I found an alternative Query here: 
https://stackoverflow.com/a/16996103

And edited the File as follows:
```
        /**
         * Remove artists which have no albums and no tracks
         * @return int Number of removed artists
         */
        private function removeObsoleteArtists() {
                /* 08.08.2022 Performance tuning by JR
                return $this->db->executeUpdate(
                        'DELETE FROM `*PREFIX*music_artists` WHERE `id` NOT IN (
                                SELECT `album_artist_id` FROM `*PREFIX*music_albums`
                                UNION
                                SELECT `artist_id` FROM `*PREFIX*music_tracks`
                        )'
                );
                */
                return $this->db->executeUpdate(
                        'DELETE FROM `*PREFIX*music_artists` WHERE NOT EXISTS (
                                SELECT 1
                                FROM (
                                        SELECT `album_artist_id` FROM `*PREFIX*music_albums`
                                        UNION
                                        SELECT `artist_id` FROM `*PREFIX*music_tracks`
                                ) AS existing
                                WHERE `*PREFIX*music_artists`.`id` = `album_artist_id`
                                LIMIT 1
                        )'
                );
        }
```

Hope that Helps :)

**Comment by paulijar** [comment]
Okay, thanks.

On my setup with MySQL, the query in question takes 30-60 ms. On the other hand, your proposed query takes ~500 ms, so it's about 10 times slower. In my DB, the table `oc_music_tracks` has 67073 rows, `oc_music_albums` has 2066 rows, and `oc_music_artists` has 538 rows.

How would the following alternative query perform on your system? On my setup, it's even faster than the original, taking 10-20 ms to execute. And as it still avoids the `NOT IN` with sub-query, I would assume it to perform well on PostgreSQL as well.
```
DELETE FROM `oc_music_artists` WHERE 
    NOT EXISTS (SELECT 1 FROM `oc_music_albums` WHERE `oc_music_artists`.`id` = `album_artist_id` LIMIT 1) 
    AND 
    NOT EXISTS (SELECT 1 FROM `oc_music_tracks` WHERE `oc_music_artists`.`id` = `artist_id` LIMIT 1)
```

**Comment by ravermeister** [comment]
I'll have a look tomorrow and report back✌️

**Comment by ravermeister** [comment]
Hi, 
I've got following Row counts:
| Table | Row Count |
|-------|-------------|
| oc_music_tracks | 236563 |
| oc_music_albums | 83208 |
| oc_music_artists | 69112 |

The Query you've Provided works and the **first** time I executed it took:  
`Query complete 00:00:02.568`
the **second and on going** time it took about:  
`Query complete 00:00:00.542`

The Query I've used *after your query* took: `Query complete 00:00:00.916`
so yours is slightly faster :) I'll use that for now. Thanks! :)

**Comment by paulijar** [comment]
Okay, thanks for the confirmation. You probably have query caching enabled on your DB which would explain the reduced run time on consecutive runs. And most users probably don't have that many rows in those tables which is why no-one has reported such dramatic performance issues before.

I have been planning to make the next Music app release soon and I will include the optimized query there.

**Comment by ravermeister** [comment]
> Okay, thanks.
> 
> On my setup with MySQL, the query in question takes 30-60 ms. On the other hand, your proposed query takes ~500 ms, so it's about 10 times slower. In my DB, the table `oc_music_tracks` has 67073 rows, `oc_music_albums` has 2066 rows, and `oc_music_artists` has 538 rows.
> 
> How would the following alternative query perform on your system? On my setup, it's even faster than the original, taking 10-20 ms to execute. And as it still avoids the `NOT IN` with sub-query, I would assume it to perform well on PostgreSQL as well.
> 
> ```
> DELETE FROM `oc_music_artists` WHERE 
>     NOT EXISTS (SELECT 1 FROM `oc_music_albums` WHERE `oc_music_artists`.`id` = `album_artist_id` LIMIT 1) 
>     AND 
>     NOT EXISTS (SELECT 1 FROM `oc_music_tracks` WHERE `oc_music_artists`.`id` = `artist_id` LIMIT 1)
> ```

Thinking about the query, shouldn't it be an `or not exists` to correctly reflect the Union?

**Comment by paulijar** [comment]
> Thinking about the query, shouldn't it be an or not exists to correctly reflect the Union?

Nope, because then not having an album or not having a track would become a sufficient condition to delete an artist. So we would lose all those artists which have no own albums (a featured artist on some album) and those which have no own tracks (most commonly an "artist" named "Various artists" which is used as *album artist* of some compilation album).

What probably tricks you is that we now have the `NOT` operation on different side of the parentheses. As you probably know, in boolean algebra
`not (A or B) <=> (not A) and (not B)`

**Comment by paulijar** [comment]
The optimized query is now released as part of Music v1.6.0.
